Code/GraphMol/MolInterchange/Parser.cpp

```diff
@@ -281,6 +291,286 @@ void readPartialCharges(RWMol *mol, const rj::Value &repVal,
+    res = new RecursiveStructureQuery(mol);
+  } else if (tag == MolPickler::QUERY_ATOMRING) {
+    res = new AtomRingQuery();
+    static_cast<EqualityQuery<int, Atom const *, true> *>(res)->setVal(
+        repVal["val"].GetInt());
+    if (repVal.HasMember("tol")) {
+      static_cast<EqualityQuery<int, Atom const *, true> *>(res)->setTol(
+          repVal["tol"].GetInt());
+    }
+  } else {
+    res = readBaseQuery(owner, repVal, params);
+  }
+  if (res) {
+    finishQuery(owner, res, repVal, atomDefaults, bondDefaults, params);
+  }
+  return res;
+}
+Query<int, Bond const *, true> *readQuery(Bond const *bond,
+                                          const rj::Value &repVal,
+                                          const DefaultValueCache &atomDefaults,
+                                          const DefaultValueCache &bondDefaults,
+                                          const JSONParseParameters &params) {
+  PRECONDITION(bond, "no owner");
+  if (!repVal.HasMember("tag")) {
+    throw FileParseException("Bad Format: missing bond query tag");
+  }
+  Query<int, Bond const *, true> *res = nullptr;
+  res = readBaseQuery(bond, repVal, params);
+  if (res) {
+    finishQuery(bond, res, repVal, atomDefaults, bondDefaults, params);
+  }
+
+  return res;
+}
+
+void readQueries(RWMol *mol, const rj::Value &repVal,
+                 const DefaultValueCache &atomDefaults,
+                 const DefaultValueCache &bondDefaults,
+                 const JSONParseParameters &params) {
+  PRECONDITION(mol, "no molecule");
+  PRECONDITION(repVal["name"].GetString() == std::string("rdkitQueries"),
+               "bad queries");
+  if (!repVal.HasMember("formatVersion")) {
+    throw FileParseException("Bad Format: missing format_version");
+  }
+  if (repVal["formatVersion"].GetInt() > 10) {
+    BOOST_LOG(rdWarningLog) << "RDKit query representation format version "
+                            << repVal["formatVersion"].GetInt()
+                            << " too recent. Ignoring it." << std::endl;
+    return;
+  }
+  {
+    const auto &miter = repVal.FindMember("atomQueries");
+    if (miter != repVal.MemberEnd()) {
+      size_t idx = 0;
+      for (const auto &val : miter->value.GetArray()) {
+        if (!val.IsObject()) {
+          throw FileParseException("Bad Format: atomQuery not object");
+        }
+        if (!val.HasMember("tag")) {
+          // nothing here, continue
+          continue;
+        }
+        if (idx >= mol->getNumAtoms()) {
+          throw FileParseException("too much atom data found");
+        }
+        auto atom = mol->getAtomWithIdx(idx);
+        CHECK_INVARIANT(atom != nullptr, "no atom");
+        // we need to replace the current atom with a query atom:
+        QueryAtom qatom(*atom);
+        // that copy created a bunch of query info by default,
+        // but we want to get the info from the JSON, so delete
+        // that:
+        qatom.setQuery(nullptr);
```

Comment:
You always call setQuery, does it really need to be set to nullptr here?

Reply:
(It probably saves when you copy the atom into the molecule...)

Reply:
yeah, it saves the additional copy of the query info which we'll throw away anyway when copying the atom into the mol